content/recipes/authoring/kata-snippets/full-tests.md

```diff
@@ -5,6 +5,82 @@ prev: /recipes/authoring/kata-snippets/sample-tests/
+- **Use appropriate random utilities available in your language**. Know how to use RNGs, how to randomly generate types of inputs you need, be it numbers in a range, large numbers, strings, big integers, floating point values, collections.  Know how to do random sampling of collections, how to shuffle them, how to avoid duplicates. See this [obligatory XKCD comic](https://xkcd.com/221/) for how NOT to do random tests.
+- **Random tests shouldn't be grouped by output type or behavior**, especially if the expected output is a boolean variable (e.g. checking that some input satisfies some criteria), or when it comes to error checking (solution throwing an exception in some specific situations). Hence, the order of tested scenarios should be unpredictable too. One possible way to achieve this is to generate and collect a set of random inputs for all required scenarios and shuffle them before the actual testing.
+- Under some rare circumstances, it's allowed to use so called [**randomized tests**](/glossary/#randomized-tests) instead of fully random ones. For some types of problems (for example, kata simulating chess, or problems with small set of possible inputs) it may be too complex or infeasible to generate inputs randomly. In such rare cases, it's accepted to use a predefined, hard-coded set of inputs (and possibly expected outputs too). Before the tests are run, the set of inputs should be shuffled, and eventually randomly sampled at each run to make the hard-coding of the results more tedious for the user. If possible, some additional transformation can be applied randomly (for example array can be reversed, game board can be rotated, sides can be flipped) if it can be easily accounted for in the result of the test.
+- **Random tests should be run after fixed tests.** Not all testing frameworks allow for easy ordering of tests, but fixed tests should run, and eventually fail, before random tests.
+- **Difficulty should be consistent between test runs.** When randomized ranges are very large, it becomes possible for some users to receive many small, easy inputs while other users receive the exact opposite. Make sure that your random tests are built in a way which minimizes the chances for such a situation to occur. If you want to test difficult inputs, split the test cases into a set of easy ones and a set of difficult ones, and test them separately.
+- **Make debugging of random tests easy.** Sometimes, when random tests fail, they are very difficult to debug because the input is very large and cannot be easily read. If necessary, split your random tests into two batches: one with small, debuggable inputs, and the other one with proper, large input values. Note that both parts should still contain all applicable scenarios.
```

Comment:
Unless a kata relies mainly on the random tests by design, random tests shouldn't fail if no fixed tests failed, so debugging with random tests shouldn't be a thing in most cases, thus their output isn't too important.

Reply:
```suggestion
- If it's not possible to build fixed tests in a way they'd be able to catch most of the failures and you have to rely on random tests to catch errors, **make debugging of random tests easy.** Sometimes, when random tests fail, they are very difficult to debug because the input is very large and cannot be easily read. If necessary, split your random tests into two batches: one with small, debuggable inputs, and the other one with proper, large input values. Note that both parts should still contain all applicable scenarios.
```

Reply:
@error256 yes but in some cases, it's pretty hard to actually figure out possible edge cases, because they will depend on the implemetation of the user. Alphametic solver, for example. You can imagine a lot of special cases up front, but considering the length of the code, it's impossible to ensure that you're covering everything. So a first batch of small random tests is still of interest (but not an excuse to not put as much egde cases as possible in the fixed tests).

Reply:
Yes, but it can be considered a kata that relies on random testing.
Anyway, what's the preferred workflow here? Should individual suggestions be committed or it's better to batch-commit there in a group with other changes?

Reply:
For typos, we usually commit suggestions with fixes in batch. For paragraphs, we can commit single ones after we think they reached good state. If you see any suggestion addressing your remark and you think it's good, you can commit it without waiting for others.

You can also write your own suggestions or edit other suggestions to fix typos or wording.